src/Functions/FunctionVariantAdaptor.cpp

```diff
@@ -0,0 +1,746 @@
+                if (result_variants[j]->equals(*variant_result_type))
+                {
+                    result_discriminators[i] = j;
+                    break;
+                }
+            }
+        }
+
+        if (!result_discriminators[i])
+            throw Exception(
+                ErrorCodes::LOGICAL_ERROR,
+                "Cannot find variant type {} in result Variant type {} during execution of {}",
+                variant_result_type->getName(),
+                result_type->getName(),
+                getName());
+    }
+
+    /// Set variant columns in the result
+    for (size_t i = 1; i < variants_results.size(); ++i)
+    {
+        if (variants_results[i] && result_discriminators[i])
+        {
+            auto global_discr = *result_discriminators[i];
+            auto local_discr = result_variant.localDiscriminatorByGlobal(global_discr);
+            result_variant.getVariantPtrByLocalDiscriminator(local_discr) = variants_results[i];
+        }
+    }
+
+    /// Build discriminators and offsets
+    auto & result_discriminators_col = result_variant.getLocalDiscriminators();
+    auto & result_offsets = result_variant.getOffsets();
+    result_discriminators_col.reserve(variant_column.size());
+    result_offsets.reserve(variant_column.size());
+
+    for (size_t i = 0; i != selector.size(); ++i)
+    {
+        if (selector[i] == 0 || !variants_results[selector[i]])
+        {
+            result_discriminators_col.push_back(ColumnVariant::NULL_DISCRIMINATOR);
+            result_offsets.emplace_back();
+        }
+        else
+        {
+            auto global_discr = *result_discriminators[selector[i]];
+            auto local_discr = result_variant.localDiscriminatorByGlobal(global_discr);
+            result_discriminators_col.push_back(local_discr);
+            result_offsets.push_back(variants_offsets[i]);
+        }
+    }
+
+    return result;
+}
+
+ColumnPtr ExecutableFunctionVariantAdaptor::executeImpl(
+    const ColumnsWithTypeAndName & arguments, const DataTypePtr & result_type, size_t input_rows_count) const
+{
+    return executeImpl(arguments, result_type, input_rows_count, false);
+}
+
+ColumnPtr ExecutableFunctionVariantAdaptor::executeDryRunImpl(
+    const ColumnsWithTypeAndName & arguments, const DataTypePtr & result_type, size_t input_rows_count) const
+{
+    return executeImpl(arguments, result_type, input_rows_count, true);
+}
+
+FunctionBaseVariantAdaptor::FunctionBaseVariantAdaptor(
+    std::shared_ptr<const IFunctionOverloadResolver> function_overload_resolver_, DataTypes arguments_)
+    : function_overload_resolver(function_overload_resolver_)
+    , arguments(arguments_)
+{
+    std::optional<size_t> first_variant_index;
+    for (size_t i = 0; i != arguments.size(); ++i)
+    {
+        if (isVariant(arguments[i]))
+        {
+            if (!first_variant_index.has_value())
+            {
+                first_variant_index = i;
+                break;
+            }
```

Comment:
```suggestion
            first_variant_index = i;
            break;
```

This `if` seems not needed as we break on first Variant argument